### PR TITLE
RR-468 cypress tests for the pagination controls

### PR DESCRIPTION
**Pull request by chrimesdev**
## Description

Cypress tests for the pagination controls

https://dsdmoj.atlassian.net/browse/RR-468

### CODE REVIEW

File: integration_tests/pages/prisonerList/PrisonerListPage.ts [hunk 1]
```diff
@@ -76,6 +76,33 @@ export default class PrisonerListPage extends Page {
...
+
+  hasNextLinkDisplayed(): PrisonerListPage {
+    this.paginationControls().find('li:last-of-type a').should('contain', 'Next')
+    return this
+  }
+
+  hasNumberOfPagesDisplayed(expected: number): PrisonerListPage {
+    this.paginationControls()
+      .find('li.moj-pagination__item')
+      .should('have.length', expected + 1)
```
Comment (chrimesdev) [inline integration_tests/pages/prisonerList/PrisonerListPage.ts:102]: Need to improve this - will fix now

File: integration_tests/pages/prisonerList/PrisonerListPage.ts [hunk 2]
```diff
@@ -76,6 +76,35 @@ export default class PrisonerListPage extends Page {
...
+
+  hasNextLinkDisplayed(): PrisonerListPage {
+    this.paginationControls().find('li:last-of-type a').should('contain', 'Next')
+    return this
+  }
+
+  hasNumberOfPagesDisplayed(expected: number): PrisonerListPage {
+    this.paginationControls()
+      .find('li.moj-pagination__item')
+      .not('.moj-pagination__item--next')
```
Comment (chrimesdev) [inline integration_tests/pages/prisonerList/PrisonerListPage.ts:102]: Maybe not the best approach but only way I could think of without having the ability to pass attributes in the Nunjucks macros!

Comment (nathanrussell-moj-digital) [reply integration_tests/pages/prisonerList/PrisonerListPage.ts:102]: It's fine; I agree, we don't want to litter the nunjucks markup with too many attributes etc just to support tests.

File: integration_tests/pages/prisonerList/PrisonerListPage.ts [hunk 3]
```diff
@@ -76,6 +76,35 @@ export default class PrisonerListPage extends Page {
...
+    return this
+  }
+
+  setPaginationNumber(value: number): PrisonerListPage {
+    this.paginationControls().find(`li:nth-of-type(${value}) a`).click()
+    return this
+  }
+
+  hasPreviousLinkDisplayed(): PrisonerListPage {
+    this.paginationControls().find('li:first-of-type a').should('contain', 'Previous')
```
Comment (nathanrussell-moj-digital) [inline integration_tests/pages/prisonerList/PrisonerListPage.ts:90]: Could you use your `paginationFirstLink` method here?
```
this.paginationFirstLink().should('contain', 'Previous')
```

Would that work? And then do the same thing for the next link displayed?

File: integration_tests/e2e/prisonerList/prisonerList.cy.ts [hunk 4]
```diff
@@ -148,6 +148,40 @@ context(`Display the prisoner list screen`, () => {
...
+
+      // Then
+      prisonerListPage //
+        .hasPaginationControlsDisplayed()
+        .hasNumberOfPagesDisplayed(10)
+        .setPaginationNumber(7)
+        .hasPreviousLinkDisplayed()
+        .hasNextLinkDisplayed()
+    })
+  })
```
Comment (nathanrussell-moj-digital) [inline integration_tests/e2e/prisonerList/prisonerList.cy.ts:184]: I'm not sure this is what I had in mind .... I think I was thinking more something like this (most of it is pseudo code)
```
    it('should display pagination controls, displaying 2 to 11 with the next and previous links, on the 7th page of the prisoner list', () => {
      // Given
      cy.signIn()
      cy.visit('/')
      const prisonerListPage = Page.verifyOnPage(PrisonerListPage)

      // When
      prisonerListPage.gotoPage(7) // I think "gotoPage" is a better name than your current "setPaginationNumber" method, but its basically the same thing

      // Then
      prisonerListPage //
        .doesNotHavePaginationLinkForPage(1)
        .hasPreviousLinkDisplayed()
        .hasPaginationLinkForPage(2)
        .hasPaginationLinkForPage(3)
        .hasPaginationLinkForPage(4)
        .hasPaginationLinkForPage(5)
        .hasPaginationLinkForPage(6)
        .paginationCurrentPageIs(7)
        .hasPaginationLinkForPage(8)
        .hasPaginationLinkForPage(9)
        .hasPaginationLinkForPage(10)
        .hasPaginationLinkForPage(11)
        .hasPaginationLinkForPage(12)
        .hasNextLinkDisplayed()
        .doesNotHavePaginationLinkForPage(13)
    })
```
Something like that perhaps?
There might be better ways to express that we are expecting a specific set of pagination controls,  but you get the idea

Comment (chrimesdev) [reply integration_tests/e2e/prisonerList/prisonerList.cy.ts:184]: Will revisit!

File: integration_tests/pages/prisonerList/PrisonerListPage.ts (hunk 3 above)
Comment (chrimesdev) [reply integration_tests/pages/prisonerList/PrisonerListPage.ts:90]: Good spot, just realised that `paginationFirstLink` isn't even used currently - maybe I started this trail of thought and got lost!

**Review by nathanrussell-moj-digital — APPROVED** [review]
LGTM

**Pull request merged**